src/_utils/focusVisible/FocusVisibleProvider.tsx

```diff
@@ -0,0 +1,131 @@
+    document.removeEventListener('mousemove', onInitialPointerMove)
+    document.removeEventListener('mousedown', onInitialPointerMove)
+    document.removeEventListener('mouseup', onInitialPointerMove)
+    document.removeEventListener('pointermove', onInitialPointerMove)
+    document.removeEventListener('pointerdown', onInitialPointerMove)
+    document.removeEventListener('pointerup', onInitialPointerMove)
+    document.removeEventListener('touchmove', onInitialPointerMove)
+    document.removeEventListener('touchstart', onInitialPointerMove)
+    document.removeEventListener('touchend', onInitialPointerMove)
+  }
+
+  /**
+   * If the most recent user interaction was via the keyboard;
+   * and the key press did not include a meta, alt/option, or control key;
+   * then the modality is keyboard. Otherwise, the modality is not keyboard.
+   * Apply `focus-visible` to any current active element and keep track
+   * of our keyboard modality state with `hadKeyboardEvent`.
+   * @param {KeyboardEvent} e
+   */
+  const onKeyDown = (e: KeyboardEvent) => {
+    if (e.metaKey || e.altKey || e.ctrlKey) {
+      return
+    }
+
+    setHadKeyboardEvent(true)
+  }
+  /**
+   * If the user changes tabs, keep track of whether or not the previously
+   * focused element had .focus-visible.
+   */
+  const onVisibilityChange = () => {
+    if (document.visibilityState === 'hidden') {
+      // If the tab becomes active again, the browser will handle calling focus
+      // on the element (Safari actually calls it twice).
+      // If this tab change caused a blur on an element with focus-visible,
+      // re-apply the class when the user switches back to the tab.
+      setHadKeyboardEvent(true)
+      addInitialPointerMoveListeners()
+    }
+  }
+
+  useEffect(() => {
+    // For some kinds of state, we are interested in changes at the global scope
+    // only. For example, global pointer input, global key presses and global
+    // visibility change should affect the state at every scope:
+    document.addEventListener('keydown', onKeyDown, true)
+    document.addEventListener('mousedown', onPointerDown, true)
+    document.addEventListener('pointerdown', onPointerDown, true)
+    document.addEventListener('touchstart', onPointerDown, true)
+    document.addEventListener('visibilitychange', onVisibilityChange, true)
+
+    addInitialPointerMoveListeners()
+
+    return () => {
+      document.removeEventListener('keydown', onKeyDown, true)
+      document.removeEventListener('mousedown', onPointerDown, true)
+      document.removeEventListener('pointerdown', onPointerDown, true)
+      document.removeEventListener('touchstart', onPointerDown, true)
+      document.removeEventListener('visibilitychange', onVisibilityChange, true)
+
+      removeInitialPointerMoveListeners()
+    }
+  }, [setHadKeyboardEvent])
+
+  return (
+    <FocusVisibleContext.Provider value={{ hadKeyboardEvent, isInitialized: true }}>
```

Comment:
`hadKeyboardEvent ` updated from this component state so might have no desired re-render effect.
It will change when user switch form keyboard navigation to pointer navigation.